# DATA PROCESSING
****

## Recognizing, describing and cleaning the data

###### i. First things first: Import libraries of interest.

In [1]:
import pandas as pd
import numpy as np

###### ii. Useful documentation:
* <a href="https://www.earthdatascience.org/courses/intro-to-earth-data-science/file-formats/use-text-files/format-text-with-markdown-jupyter-notebook/" target="_blank">Format Text In Jupyter Notebook</a>
* <a href="https://stackoverflow.com/questions/8924173/how-do-i-print-bold-text-in-python" target="_blank">How do I print bold text in Python?</a>
* <a href="https://www.nrcs.usda.gov/wps/portal/nrcs/detail/ny/home/?cid=nrcs143_013696" target="_blank">Every State FIPS code</a>
* <a href="https://www.nrcs.usda.gov/wps/portal/nrcs/detail/national/home/?cid=nrcs143_013697" target="_blank">Every County FIPS code</a>


****
### I. Chemicals Data Set

In [2]:
#Load Chemicals and preview
df_chemicals = pd.read_csv("datasets/chemicals.csv")
df_chemicals.head(3).append(df_chemicals.tail(3))

cws_name chemical_species  \
0         CALIFORNIA WATER SERVICE - LIVERMORE          Uranium   
1                            CITY OF LIVERMORE          Uranium   
2                           CITY OF PLEASANTON          Uranium   
882316                 PORT EDWARDS WATERWORKS   Trihalomethane   
882317                       VESPER WATERWORKS   Trihalomethane   
882318  WIS RAPIDS WATER WORKS & LIGHTING COMM   Trihalomethane   

             contaminant_level   county     pws_id  pop_served       state  \
0       Less than or equal MCL  Alameda  CA0110003       57200  California   
1       Less than or equal MCL  Alameda  CA0110011       26400  California   
2       Less than or equal MCL  Alameda  CA0110008       70600  California   
882316  Less than or equal MCL     Wood  WI7720113        1866   Wisconsin   
882317  Less than or equal MCL     Wood  WI7720134         641   Wisconsin   
882318  Less than or equal MCL     Wood  WI7720108       20000   Wisconsin   

       unit_measurement  value  year   fips  state_fips  
0          micrograms/L   1.54  2000   6001           6  
1          micrograms/L   1.64  2000   6001           6  
2          micrograms/L   1.64  2000   6001           6  
882316     micrograms/L   4.70  2016  55141          55  
882317     micrograms/L  68.90  2016  55141          55  
882318     micrograms/L  53.46  2016  55141          55

#### First Impressions:
* Probably won't be using the name of the Water systems per County.
* Must look through each String Variable
* Must look for coherence between County-Fips and State-State Fips.

****
### a) Recognizing variables and checking for null observations

In [3]:
#Recognition of the variables available
print(df_chemicals.info())
print()

#Null observations checking.
print(df_chemicals.isnull().any())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 882319 entries, 0 to 882318
Data columns (total 12 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   cws_name           882319 non-null  object 
 1   chemical_species   882319 non-null  object 
 2   contaminant_level  882319 non-null  object 
 3   county             882319 non-null  object 
 4   pws_id             882319 non-null  object 
 5   pop_served         882319 non-null  int64  
 6   state              882319 non-null  object 
 7   unit_measurement   882319 non-null  object 
 8   value              882319 non-null  float64
 9   year               882319 non-null  int64  
 10  fips               882319 non-null  int64  
 11  state_fips         882319 non-null  int64  
dtypes: float64(1), int64(4), object(7)
memory usage: 80.8+ MB
None

cws_name             False
chemical_species     False
contaminant_level    False
county               False
pws_id               False

##### Conclusions:
* Every record is complete.
* There's not null data we need to handle.
* We must know what information is present inside each variable.
****

### b) Recognizing observations per variable of interest

#### VARIABLE TYPE: STRING

#### 1. 'chemical_species'

In [4]:
list(df_chemicals['chemical_species'].unique())

['Uranium',
 'Arsenic',
 'DEHP',
 'Nitrates',
 'Halo-Acetic Acid',
 'Trihalomethane']

##### Conclusions:
* The span of our data is 6 contaminants we don't know much about.

***
<ul><h5> General description and industry-usage of this chemicals</h5></ul>

<b>Uranium:</b> (<a href="https://www.who.int/ionizing_radiation/pub_meet/en/Depluranium2.pdf?ua=1" target="_blank">source</a>):</ul></b>
* Uranium is weakly radioactive because all isotopes of uranium are unstable

<ul><b>Industry Uses:</b></ul>

* Radiation shielding
*  Counterbalance weights and ballast Vessels and equipment, such as boats and satellites require a large amount of weight to be carried in the form of ballast. 
* Military Uses Depleted uranium (and associated uranium-titanium alloys [typically 0.75 wt % Ti]) have been employed by the military as a component of heavy tank armour and armour-piercing munitions
* Nuclear power plants.

<b>Arsenic:</b> (<a href="https://www.who.int/news-room/fact-sheets/detail/arsenic#:~:text=Arsenic%20is%20naturally%20present%20at,toxic%20in%20its%20inorganic%20form.&text=Long%2Dterm%20exposure%20to%20arsenic,with%20cardiovascular%20disease%20and%20diabetes" target="_blank">source</a>):</ul></b>

* Arsenic is naturally present at high levels in the groundwater of a number of countries.
* Arsenic is highly toxic in its inorganic form.
* Long-term exposure to arsenic from drinking-water and food can cause cancer and skin lesions. It has also been associated with cardiovascular disease and diabetes. In utero and early childhood exposure has been linked to negative impacts on cognitive development and increased deaths in young adults.

<ul><b>Industry Uses:</b></ul>

* Arsenic is a natural component of the earth’s crust and is widely distributed throughout the environment in the air, water and land
* Arsenic is used industrially as an alloying agent, as well as in the processing of glass, pigments, textiles, paper, metal adhesives, wood preservatives and ammunition. 
* Arsenic is also used in the hide tanning process and, to a limited extent, in pesticides, feed additives and pharmaceuticals.
* People who smoke tobacco can also be exposed to the natural inorganic arsenic content of tobacco because tobacco plants can take up arsenic naturally present in the soil. Also, in the past, the potential for elevated arsenic exposure was much greater when tobacco plants used to be treated with lead arsenate insecticide

<b>DEHP (<a href="https://en.wikipedia.org/wiki/Bis(2-ethylhexyl)_phthalate" target="_blank">source</a>):</ul></b>
* Organic compound
* Colorless viscous liquid soluble in oil, but not in water.

<ul><b>Industry Uses:</b></ul>

* Due to its suitable properties and the low cost, DEHP is widely used as a plasticizer in manufacturing of articles made of *PVC*. *PVC* is used extensively in *sewage pipe* due to its low cost, chemical resistance, and ease of jointing.
* *Plastics* may contain 1% to 40% of DEHP.
* Used as a *hydraulic fluid* and as a dielectric fluid in *capacitors*.
* Use as a solvent in *glowsticks*.
* DEHP is a component of *many household items*, including tablecloths, floor tiles, shower curtains, garden hoses, rainwear, dolls, toys, shoes, medical tubing, furniture upholstery, and swimming pool liners.
*  Common exposures come from the use of DEHP as a fragrance carrier in cosmetics, personal care products, laundry detergents, colognes, scented candles, and air fresheners.
* **DEHP can leach into drinking water from discharges from rubber and chemical factories**


<b>Nitrates:</b> (<a href="https://www.health.state.mn.us/communities/environment/water/contaminants/nitrate.html#MinnesotaWater" target="_blank">source</a>):</ul></b>
* Nitrate is a compound that naturally occurs and has many human-made sources.

<ul><b>Industry Uses:</b></ul>

* It is often difficult to pinpoint sources of nitrates because there are so many possibilities.
* Sources of nitrogen and nitrates may include runoff or seepage from fertilized agricultural lands, municipal and industrial waste water, refuse dumps, animal feedlots, septic tanks and private sewage disposal systems, urban drainage and decaying plant debris. 
* Geologic formations and direction of ground water flow also may influence nitrate concentration.

<b>Halo-Acetic Acid:</b> (<a href="https://www.simcoemuskokahealth.org/Topics/SafeWater/drinkingwater/chemicalsminerals/Haloacetic-Acids-in-Drinking-Water#:~:text=Haloacetic%20acids%20(HAAs)%20are%20a,matter%20in%20the%20source%20water." target="_blank">source</a>):</ul></b>
* Haloacetic acids (HAAs) are a group of compounds that can form in the water distribution systems when chlorine used to disinfect drinking water reacts with naturally occurring organic matter in the source water

<ul><b>Industry Uses:</b></ul>

* It is often difficult to pinpoint sources of nitrates because there are so many possibilities.
* Sources of nitrogen and nitrates may include runoff or seepage from fertilized agricultural lands, municipal and industrial waste water, refuse dumps, animal feedlots, septic tanks and private sewage disposal systems, urban drainage and decaying plant debris. 
* Geologic formations and direction of ground water flow also may influence nitrate concentration.

<b>Trihalomethane:</b> </b> (<a href="https://www.simcoemuskokahealth.org/Topics/SafeWater/drinkingwater/chemicalsminerals/Haloacetic-Acids-in-Drinking-Water#:~:text=Haloacetic%20acids%20(HAAs)%20are%20a,matter%20in%20the%20source%20water." target="_blank">source</a>):</ul></b>
* Trihalomethanes (THMs) are chemical compounds in which three of the four hydrogen atoms of methane (CH4) are replaced by halogen atoms.
* Many trihalomethanes find uses in industry as solvents or refrigerants.

<ul><b>Industry Uses:</b></ul>

* Trifluoromethane and chlorodifluoromethane are both used as refrigerants. Trihalomethanes released to the environment break down faster than chlorofluorocarbons (CFCs), thereby doing much less damage to the ozone layer.
* Chloroform is a common solvent in organic chemistry.
* Trihalomethanes are formed as a by-product predominantly when chlorine is used to disinfect drinking water. They are generally referred to as disinfection by-products.

***
#### 2. 'contaminant_level'

In [5]:
list(df_chemicals['contaminant_level'].unique())

['Less than or equal MCL', 'Non Detect', 'Greater than MCL']

##### Conclusions:
* Span of 3 possible categories.
* *MCL* stands for *Maximum Contaminant Level*

***
#### 3. 'state' and 'state_fips'

In [6]:
#Have the names of the States a unique 'state_fips' code?
len_county=len(df_chemicals['state'].unique())
len_fips=len(df_chemicals['state_fips'].unique())
print('\033[1m We have {0:.0f} State names and {1:.0f} State codes for those names.'.format(len_county,len_fips))

 We have 27 State names and 27 State codes for those names.


##### Conclusions:
* We have 27 county names and 27 county codes for those names
* We can use the name or the fips code for States.

***
#### 4. 'county' and 'fips'

In [7]:
#Have the names of the Counties a unique 'fips' code?
len_county=len(df_chemicals['county'].unique())
len_fips=len(df_chemicals['fips'].unique())
print('\033[1m *We have {0:.0f} County names and {1:.0f} County codes for those names.\n'.format(len_county,len_fips))
print('\033[1m **Are there County names that have more than 1 fips code?')

 *We have 1045 County names and 1500 County codes for those names.

 **Are there County names that have more than 1 fips code?


In [8]:
#County and State data in one place
df_county_fips=pd.concat([df_chemicals['county'],df_chemicals['fips'],df_chemicals['state'],df_chemicals['state_fips']],axis=1).drop_duplicates()
df_county_fips

county   fips       state  state_fips
0         Alameda   6001  California           6
89         Alpine   6003  California           6
101        Amador   6005  California           6
140         Butte   6007  California           6
329     Calaveras   6009  California           6
...           ...    ...         ...         ...
639435  Van Buren  26159    Michigan          26
639466  Washtenaw  26161    Michigan          26
639541      Wayne  26163    Michigan          26
639756    Wexford  26165    Michigan          26
651627      Ralls  29173    Missouri          29

[1500 rows x 4 columns]

In [9]:
dup_counties=df_county_fips.groupby('county').agg({'fips':'count'})
dup_counties.rename(columns={'fips':'n_fips_dupl'},inplace=True)
print(dup_counties[dup_counties.n_fips_dupl>1])
print()
print('\033[1m *Yes! There are names of Counties that have more than 1 fips code.\n')
print('\033[1m **Is it a mistake or could there be Counties with the same name in different States?')
print('\033[1m   Lets check top and bottom 2.')

           n_fips_dupl
county                
Adair                3
Adams                5
Allegany             2
Allen                3
Anderson             3
...                ...
Worcester            2
Worth                2
Wright               3
Wyoming              3
York                 3

[190 rows x 1 columns]

 *Yes! There are names of Counties that have more than 1 fips code.

 **Is it a mistake or could there be Counties with the same name in different States?
   Lets check top and bottom 2.


In [10]:
df_county_fips[df_county_fips.county=='Adair']

county   fips     state  state_fips
17923   Adair  19001      Iowa          19
18760   Adair  21001  Kentucky          21
100892  Adair  29001  Missouri          29

In [11]:
df_county_fips[df_county_fips.county=='Adams']

county   fips         state  state_fips
12361   Adams   8001      Colorado           8
32167   Adams  42001  Pennsylvania          42
37253   Adams  55001     Wisconsin          55
74529   Adams  19003          Iowa          19
160600  Adams  53001    Washington          53

In [12]:
df_county_fips[df_county_fips.county=='Wyoming']

county   fips          state  state_fips
28424  Wyoming  36121       New York          36
35628  Wyoming  42131   Pennsylvania          42
37252  Wyoming  54109  West Virginia          54

In [13]:
df_county_fips[df_county_fips.county=='York']

county   fips           state  state_fips
35692   York  42133    Pennsylvania          42
36149   York  45091  South Carolina          45
89965   York  23031           Maine          23

###### Anyway, let's make sure that no name is repeated in a single State.

In [14]:
df_county_fips=pd.merge(df_county_fips, dup_counties, how='left', left_on='county', right_on='county')
dup_counties_states=df_county_fips.groupby(['county','state']).size()
dup_counties_states.rename('n_state_fips_dupl',inplace=True)
print("\033[1m *The maximum ammount of times a name of a County belongs to a specific State, in our dataset, is {0:.0f}.".format(dup_counties_states.max()))

 *The maximum ammount of times a name of a County belongs to a specific State, in our dataset, is 1.


##### Conclusions:
* Fips codes are unique.
* There are names of Counties (190) that have more than 1 fips code.
* A County name never repeats within a State.
* We must use the Counties' code (fips), when studying the information with Counties as a Factor variable.

***
#### 5. 'cws_name' and 'pws_id'

##### Conclusions:
The scope of our study won't be needing this information. For that reason we won't be checking this, as with the Counties.

***
#### 6. 'unit_measurement'

In [15]:
list(df_chemicals['unit_measurement'].unique())

['micrograms/L', 'milligrams/L']

#### Conclusions:
* Span of 2 possible categories.
* Similar nomenclature to be aware when concluding, because *Milli* **>** *Micro*:
    * **Micro=** 1*10^-6
    * **Milli=** 1*10^-3

***
#### VARIABLE TYPE: NUMERIC (INTEGER, FLOAT)

#### 7. 'pop_served'

In [16]:
print(df_chemicals['pop_served'].describe().astype('int64'))

count     882319
mean       10731
std        95200
min            0
25%          118
50%          485
75%         3030
max      8271000
Name: pop_served, dtype: int64


#### 8. 'value'

In [17]:
print(df_chemicals['value'].describe())

count    882319.000000
mean          7.434686
std          23.677866
min           0.000000
25%           0.300000
50%           1.000000
75%           4.600000
max        4527.000000
Name: value, dtype: float64


#### 9. 'year'

In [18]:
print(df_chemicals['year'].describe().astype('int64'))

count    882319
mean       2008
std           4
min        1999
25%        2005
50%        2009
75%        2013
max        2016
Name: year, dtype: int64


In [19]:
df_chemicals['year'].value_counts()

2014    68243
2013    65423
2012    62174
2015    61645
2010    59554
2007    57030
2011    55248
2016    55091
2009    54554
2006    53995
2004    52937
2008    49582
2005    48128
2003    39077
2002    33244
2000    25800
2001    22564
1999    18030
Name: year, dtype: int64

****
<h3>Earnings Data Set</h3>

In [20]:
df_earnings = pd.read_csv("datasets/earnings.csv", encoding='latin-1')
df_earnings.head(3).append(df_earnings.tail(3))

geo_id   fips                    county  total_med  \
0      0500000US01001   1001   Autauga County, Alabama      31811   
1      0500000US01003   1003   Baldwin County, Alabama      30099   
2      0500000US01005   1005   Barbour County, Alabama      23504   
21995  0500000US56041  56041     Uinta County, Wyoming      30106   
21996  0500000US56043  56043  Washakie County, Wyoming      32336   
21997  0500000US56045  56045    Weston County, Wyoming      41058   

      total_agri_fish_mine agri_fish_hunt mining_quarrying_oilgas_extract  \
0                    31793          31957                           22357   
1                    33173          27342                           80938   
2                    16167          14464                           27500   
21995                82877          34063                           86217   
21996                29844          25139                           54773   
21997                70375          23173                           75417   

      construction manufacturing wholesale_trade  ... admin_sup  \
0            30347         44452           38203  ...     31250   
1            30068         41917           43444  ...     23910   
2            20305         27630           18777  ...     20179   
21995        23618         61607           55461  ...     13722   
21996        38902         38664           36759  ...     33365   
21997        42650         42102           75000  ...     40833   

      total_edu_health_social edu_serv health_social total_arts_ent_acc_food  \
0                     30306.0    34358         26839                   11231   
1                     34506.0    37341         30968                   14924   
2                     22398.0    42264         20434                    6473   
21995                 27728.0    26211         28207                    8277   
21996                 27305.0    46667         22335                    9211   
21997                 23462.0    23510         22440                   20104   

      arts_ent_rec acc_food_serv other_ser pub_admin  year  
0            10272         11430     26279     46858  2010  
1            17414         14765     21440     38629  2010  
2            10724          5972     16090     30871  2010  
21995         5656          9177     31136     44052  2016  
21996         7375          9866     20913     42708  2016  
21997        33750         14728     23125     40588  2016  

[6 rows x 31 columns]

#### First Impressions:
* `'county'` here has the County and the State in the same variable as a String. Apparently here we have more Counties than in `Chemicals`, so it could be useful to have the State in a different variable. 
    * **Note:** We don't care about State fips because every state name have its own state fips code. If needed we have all the State (and County) fips codes to merge here
* Must look through each String Variable
* We won't be looking into coherence between county and fips code, because we already have insights in this. Because it's a lot of information we must asume each record is ok.
* We must describe each tag on Industry.

##### Variable Description:
* **geo_id:** ID with the geolocation of each County. Usefull for graph maps.
* **fips:** County Code
* **county:** County and State name
* **total_med:** Total median earnings per County
* **total_agri_fish_mine:** Subtotal median earnings in agriculture, fishing and mining
* **agri_fish_hunt:** Agricultural, Fishing and Hunting median earnings
* **mining_quarrying_oilgas_extract:** Mining, Quarrying, Oil & Gas and other extractive Industries median earnings
* **construction:** Construction Industry median earnings
* **manufacturing:** Manufacturing median earnings
* **wholesale_trade:** Wholesale Trading median earnings
* **retail_trade:** Retail Trading median earnings
* **transport_warehouse_utilities:** Transportation, Warehousing and Utilities median earnings
* **transport_warehouse:** Transport Warehouse median earnings
* **utilities:** Utilities Industry median earnings
* **information:** Information Industry median earnings
* **fin_ins_realest:** Finance and Real Estate Industries median earnings
* **fin_ins:** Financial Industry median earnings
* **realest_rent_lease:** Real Estate Renting and Leasing Industry median earnings
* **total_prof_sci_mgmt_admin:** Subtotal median earnings in proffesional, scientific, management and administrative industries
* **prof_sci_tech:** Proffesional, Scientific and Technology median earnings
* **mgmt:** Management Industry median earnings
* **admin_sup:** Administrative median earnings
* **total_edu_health_social:** Subtotal median earnings in Educational, Healt and Social Industries
* **edu_serv:** Educational median earnings
* **health_social:** Health Industry median earnings
* **total_arts_ent_acc_food:** Subtotal median earnings in Arts, Food, Public Administration, and Other Services
* **arts_ent_rec:** Art Industry median earnings
* **acc_food_serv:** Accomodation and Food median earnings
* **other_ser:** Other Services median earnings
* **pub_admin:** Public Administration median earnings
* **year:** Year of the record.

****
### a) Recognizing variables and checking for null observations

In [21]:
#Recognition of the variables available
print(df_earnings.info())
print()

#Null observations checking.
print(df_earnings.isnull().any())

#Null observations counting.
print(df_earnings.isnull().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21998 entries, 0 to 21997
Data columns (total 31 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   geo_id                           21998 non-null  object 
 1   fips                             21998 non-null  int64  
 2   county                           21998 non-null  object 
 3   total_med                        21998 non-null  int64  
 4   total_agri_fish_mine             21864 non-null  object 
 5   agri_fish_hunt                   21671 non-null  object 
 6   mining_quarrying_oilgas_extract  15751 non-null  object 
 7   construction                     21956 non-null  object 
 8   manufacturing                    21708 non-null  object 
 9   wholesale_trade                  21089 non-null  object 
 10  retail_trade                     21956 non-null  object 
 11  transport_warehouse_utilities    21928 non-null  object 
 12  transport_warehous

##### Conclusions:
* Vital variables (naming: `geo_id`, `county`, `total_med`, `year`), are complete, which is a release.
* Industries with null values are really a cero value, because those Counties had not recieved a penny from those Industries in the year of that specific record. They will be set to cero massively in the dataset.
    * This applies also to sub-totals *(Ex. total_agri_fish_mine)*.
* We must know what information is present inside each variable in general.
****

##### Massive Null Handling
Taking in mind the previous information, in this case sending every null/NaN to cero is reasonable.

In [22]:
#Recognition of the variables available
print(df_earnings.info())
print()

#Null handling
df_earnings.fillna(0, inplace = True)

#Null observations re-checking.
print(df_earnings.isnull().any())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21998 entries, 0 to 21997
Data columns (total 31 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   geo_id                           21998 non-null  object 
 1   fips                             21998 non-null  int64  
 2   county                           21998 non-null  object 
 3   total_med                        21998 non-null  int64  
 4   total_agri_fish_mine             21864 non-null  object 
 5   agri_fish_hunt                   21671 non-null  object 
 6   mining_quarrying_oilgas_extract  15751 non-null  object 
 7   construction                     21956 non-null  object 
 8   manufacturing                    21708 non-null  object 
 9   wholesale_trade                  21089 non-null  object 
 10  retail_trade                     21956 non-null  object 
 11  transport_warehouse_utilities    21928 non-null  object 
 12  transport_warehous

##### Conclusions:
* Nulls handled with success in the whole dataset

### b) Recognizing observations per variable of interest

***
#### 1. 'geo_id', 'fips' and 'county'
* They are related. 
* Because the data has no null variables and is complete we'll assume that there are not mistakes
* We'll be working with the fips code. Since we will be using this variable to merge with other datasets, everything should be fine.
* We have the data of every County and State fips code, so that will help us merging data.


***
#### 2. 'year'

In [23]:
print(df_earnings['year'].describe().astype('int64'))

count    21998
mean      2012
std          2
min       2010
25%       2011
50%       2013
75%       2015
max       2016
Name: year, dtype: int64


In [24]:
df_earnings['year'].value_counts()

2013    3143
2012    3143
2011    3143
2010    3143
2015    3142
2014    3142
2016    3142
Name: year, dtype: int64